[drogon:x64-osx] build failure

**Host Environment**
- OS: MacOS 10.15.6
- Compiler: Apple clang v. 11.0.3

**To Reproduce**
Steps to reproduce the behavior:
`./vcpkg install drogon:x64-osx`

The failure seems to happen when building mariadb.

**Failure logs** 
[106/107] : && /Library/Developer/CommandLineTools/usr/bin/clang++  -fPIC -g -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names -L/usr/local/opt/icu4c/lib drogon_ctl/CMakeFiles/drogon_ctl.dir/cmd.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_controller.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_filter.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_model.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_plugin.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_project.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_view.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/help.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/main.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/press.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/version.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/cmake.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/config.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/demoMain.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/filter_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/filter_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/gitignore.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/model_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/model_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/model_json.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/plugin_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/plugin_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_base_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_base_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_custom_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_custom_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_h.cc.o  -o drogon_ctl/drogon_ctl  libdrogon.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libtrantor.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libcares.a  -lpthread  /usr/local/vcpkg/installed/x64-osx/debug/lib/libjsoncpp.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libuuid.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libpq.dylib  /usr/local/vcpkg/installed/x64-osx/debug/lib/libmariadbclient.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libsqlite3.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libz.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libssl.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libcrypto.a  -ldl && :
FAILED: drogon_ctl/drogon_ctl 
: && /Library/Developer/CommandLineTools/usr/bin/clang++  -fPIC -g -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names -L/usr/local/opt/icu4c/lib drogon_ctl/CMakeFiles/drogon_ctl.dir/cmd.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_controller.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_filter.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_model.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_plugin.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_project.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_view.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/help.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/main.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/press.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/version.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/cmake.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/config.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/demoMain.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/filter_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/filter_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/gitignore.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/model_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/model_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/model_json.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/plugin_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/plugin_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_base_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_base_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_custom_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_custom_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_h.cc.o  -o drogon_ctl/drogon_ctl  libdrogon.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libtrantor.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libcares.a  -lpthread  /usr/local/vcpkg/installed/x64-osx/debug/lib/libjsoncpp.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libuuid.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libpq.dylib  /usr/local/vcpkg/installed/x64-osx/debug/lib/libmariadbclient.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libsqlite3.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libz.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libssl.a  /usr/local/vcpkg/installed/x64-osx/debug/lib/libcrypto.a  -ldl && :
Undefined symbols for architecture x86_64:
  "_iconv", referenced from:
      _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
  "_iconv_close", referenced from:
      _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
  "_iconv_open", referenced from:
      _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.


**Additional context**
This is a semi-common issue on OSX **if you have installed macports**. However, I do not have macports installed on my machine.

Hi @an-tao 
Thanks for your suggestion. I tried to rebuild this according to your guidelines and failed with the same error.
Then I clone a new repo to build, it still failed:

```
 Undefined symbols for architecture x86_64:
161   "_iconv", referenced from:
162       _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
163   "_iconv_close", referenced from:
164       _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
165   "_iconv_open", referenced from:
166       _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
167   "_pg_b64_dec_len", referenced from:
168       _read_server_first_message in libpq.a(fe-auth-scram.o)
169       _read_server_final_message in libpq.a(fe-auth-scram.o)
170   "_pg_b64_decode", referenced from:
171       _read_server_first_message in libpq.a(fe-auth-scram.o)
172       _read_server_final_message in libpq.a(fe-auth-scram.o)
173   "_pg_b64_enc_len", referenced from:
174       _build_client_first_message in libpq.a(fe-auth-scram.o)
175       _build_client_final_message in libpq.a(fe-auth-scram.o)
176   "_pg_b64_encode", referenced from:
177       _build_client_first_message in libpq.a(fe-auth-scram.o)
178       _build_client_final_message in libpq.a(fe-auth-scram.o)
179   "_pg_fprintf", referenced from:
180       _pqSaveParameterStatus in libpq.a(fe-exec.o)
181       _default_threadlock in libpq.a(fe-connect.o)
182       _passwordFromFile in libpq.a(fe-connect.o)
183       _defaultNoticeProcessor in libpq.a(fe-connect.o)
```

Seems we need to resolve the problems on `libmariadb `first. 
For `libmariadb`, upgrade it to the latest version 3.1.10 and add `iconv `as a feature. `drogon ` should depend on `libmariadb[iconv]` to build on Linux and OSX.

For `libpq`, the usage `find_package(PostgreSQL)`  is right. But it seems that there is lack of the linkage for `libpq `in  this part `drogon_ctl/drogon_ctl`.

```
 FAILED: drogon_ctl/drogon_ctl
159 : && /Library/Developer/CommandLineTools/usr/bin/c++  -fPIC -g -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -Wl,-search_paths_first     -Wl,-headerpad_max_install_names  drogon_ctl/CMakeFiles/drogon_ctl.dir/cmd.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create.cc.o drogon_ctl/CMakeFiles/d    rogon_ctl.dir/create_controller.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_filter.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_model.cc.o drogo    n_ctl/CMakeFiles/drogon_ctl.dir/create_plugin.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_project.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/create_v    iew.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/help.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/main.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/press.cc.o dr    ogon_ctl/CMakeFiles/drogon_ctl.dir/version.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/cmake.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/config.cc.o drogon_c    tl/CMakeFiles/drogon_ctl.dir/demoMain.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/filter_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/filter_h.cc.o drogon_    ctl/CMakeFiles/drogon_ctl.dir/gitignore.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/model_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/model_h.cc.o drogon_    ctl/CMakeFiles/drogon_ctl.dir/model_json.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/plugin_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/plugin_h.cc.o drog    on_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_base_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_base_h.cc.o drogon_ctl/CMakeFile    s/drogon_ctl.dir/restful_controller_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_custom_cc.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/r    estful_controller_custom_h.cc.o drogon_ctl/CMakeFiles/drogon_ctl.dir/restful_controller_h.cc.o  -o drogon_ctl/drogon_ctl  libdrogon.a  /Users/vcpkg/Nancy    /vcpkg/installed/x64-osx/debug/lib/libtrantor.a  /Users/vcpkg/Nancy/vcpkg/installed/x64-osx/debug/lib/libcares.a  -lpthread  /Users/vcpkg/Nancy/vcpkg/ins    talled/x64-osx/debug/lib/libjsoncpp.a  /Users/vcpkg/Nancy/vcpkg/installed/x64-osx/debug/lib/libuuid.a  /Users/vcpkg/Nancy/vcpkg/installed/x64-osx/debug/l    ib/libpq.a  /Users/vcpkg/Nancy/vcpkg/installed/x64-osx/debug/lib/libmariadbclient.a  /Users/vcpkg/Nancy/vcpkg/installed/x64-osx/debug/lib/libsqlite3.a  /    Users/vcpkg/Nancy/vcpkg/installed/x64-osx/debug/lib/libz.a  /Users/vcpkg/Nancy/vcpkg/installed/x64-osx/debug/lib/libssl.a  /Users/vcpkg/Nancy/vcpkg/insta    lled/x64-osx/debug/lib/libcrypto.a  -ldl && :
160 Undefined symbols for architecture x86_64:
161   "_iconv", referenced from:
162       _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
163   "_iconv_close", referenced from:
164       _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
165   "_iconv_open", referenced from:
166       _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
167   "_pg_b64_dec_len", referenced from:
168       _read_server_first_message in libpq.a(fe-auth-scram.o)
169       _read_server_final_message in libpq.a(fe-auth-scram.o)
170   "_pg_b64_decode", referenced from:
171       _read_server_first_message in libpq.a(fe-auth-scram.o)
172       _read_server_final_message in libpq.a(fe-auth-scram.o)
173   "_pg_b64_enc_len", referenced from:
174       _build_client_first_message in libpq.a(fe-auth-scram.o)
175       _build_client_final_message in libpq.a(fe-auth-scram.o)
176   "_pg_b64_encode", referenced from:
177       _build_client_first_message in libpq.a(fe-auth-scram.o)
178       _build_client_final_message in libpq.a(fe-auth-scram.o)
179   "_pg_fprintf", referenced from:
180       _pqSaveParameterStatus in libpq.a(fe-exec.o)
181       _default_threadlock in libpq.a(fe-connect.o)
182       _passwordFromFile in libpq.a(fe-connect.o)
183       _defaultNoticeProcessor in libpq.a(fe-connect.o)
```
Since I'm not familiar with this implemention. Maybe there is something else with this problem. We need to do further investigation about this.

Hi @an-tao 
Seems the current fix for `libpq `is not suitable. I try to remove `pg.patch` and update this part as follows:

```
  # find postgres
  find_package(pg)
  if(pg_FOUND)
    message(STATUS "libpq inc path:" ${PG_INCLUDE_DIRS})
    message(STATUS "libpq lib:" ${PG_LIBRARIES})
    target_link_libraries(${PROJECT_NAME} PRIVATE pg_lib)
...
endif(pg_FOUND)
```
as

```
 # find postgres
  find_package(PostgreSQL)
  if(PostgreSQL_FOUND)
    message(STATUS "libpq inc path:" ${PostgreSQL_INCLUDE_DIRS})
    message(STATUS "libpq lib:" ${PostgreSQL_LIBRARIES})
    target_link_libraries(${PROJECT_NAME} PRIVATE ${PostgreSQL_LIBRARIES})
...
endif(PostgreSQL_FOUND)
```
The error about `libpq ` disappered and seems only need to fix `libmariadb `now.

```
Undefined symbols for architecture x86_64:
161   "_iconv", referenced from:
162       _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
163   "_iconv_close", referenced from:
164       _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
165   "_iconv_open", referenced from:
166       _mariadb_convert_string in libmariadbclient.a(ma_charset.c.o)
167 ld: symbol(s) not found for architecture x86_64
168 clang: error: linker command failed with exit code 1 (use -v to see invocation)
169 ninja: build stopped: subcommand failed.
```
Could you please help check this again?

--- a/ports/drogon/CONTROL
+++ b/ports/drogon/CONTROL
@@ -2,4 +2,4 @@ Source: drogon
 Version: 1.0.0
 Homepage: https://github.com/an-tao/drogon
 Description:Drogon: A C++14/17 based HTTP web application framework running on Linux/macOS/Unix/Windows
-Build-Depends: trantor, zlib, jsoncpp, libmariadb, libpq, sqlite3, brotli, libuuid (!windows)
+Build-Depends: trantor, zlib, jsoncpp, libmariadb[iconv], libpq, sqlite3, brotli, libuuid (!windows)

--- a/ports/drogon/CONTROL
+++ b/ports/drogon/CONTROL
@@ -1,5 +1,6 @@
 Source: drogon
 Version: 1.0.0
+Port-Version: 1
 Homepage: https://github.com/an-tao/drogon
 Description:Drogon: A C++14/17 based HTTP web application framework running on Linux/macOS/Unix/Windows
 Build-Depends: trantor, zlib, jsoncpp, libmariadb (!osx), libmariadb[iconv] (osx), libpq, sqlite3, brotli, libuuid (!windows)

--- a/ports/drogon/CONTROL
+++ b/ports/drogon/CONTROL
@@ -2,4 +2,4 @@ Source: drogon
 Version: 1.0.0
 Homepage: https://github.com/an-tao/drogon
 Description:Drogon: A C++14/17 based HTTP web application framework running on Linux/macOS/Unix/Windows
-Build-Depends: trantor, zlib, jsoncpp, libmariadb[iconv], libpq, sqlite3, brotli, libuuid (!windows)
+Build-Depends: trantor, zlib, jsoncpp, libmariadb (!osx), libmariadb[iconv] (osx), libpq, sqlite3, brotli, libuuid (!windows)

--- a/ports/drogon/pg.patch
+++ b/ports/drogon/pg.patch
@@ -11,7 +11,7 @@ index cf3c30d..4931b93 100755
 -    message(STATUS "libpq inc path:" ${PG_INCLUDE_DIRS})
 -    message(STATUS "libpq lib:" ${PG_LIBRARIES})
 -    target_link_libraries(${PROJECT_NAME} PRIVATE pg_lib)
-+  find_package(PostgreSQL)
++  find_package(PostgreSQL REQUIRED)
 +  if(PostgreSQL_FOUND)
 +    message(STATUS "libpq inc path:" ${PostgreSQL_INCLUDE_DIRS})
 +    message(STATUS "libpq lib:" ${PostgreSQL_LIBRARIES})

--- a/ports/drogon/pg.patch
+++ b/ports/drogon/pg.patch
@@ -1,27 +1,21 @@
-diff --git a/cmake_modules/Findpg.cmake b/cmake_modules/Findpg.cmake
-index 53037f2..d13f384 100644
---- a/cmake_modules/Findpg.cmake
-+++ b/cmake_modules/Findpg.cmake
-@@ -4,20 +4,14 @@
- #
- # This module defines PG_INCLUDE_DIRS, where to find header, etc. PG_LIBRARIES,
- # the libraries needed to use PostgreSQL. pg_FOUND, If false, do not try to use
--# PostgreSQL.
--# pg_lib - The imported target library.
-+# PostgreSQL. pg_lib - The imported target library.
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index cf3c30d..4931b93 100755
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -161,11 +161,11 @@ endif(NOT WIN32)
  
- find_package(PostgreSQL)
- if(PostgreSQL_FOUND)
-   set(PG_LIBRARIES ${PostgreSQL_LIBRARIES})
-   set(PG_INCLUDE_DIRS ${PostgreSQL_INCLUDE_DIRS})
--  message(STATUS "pg inc: " ${PostgreSQL_INCLUDE_DIRS})
-   add_library(pg_lib INTERFACE IMPORTED)
--  set_target_properties(pg_lib
--                        PROPERTIES INTERFACE_INCLUDE_DIRECTORIES
--                                   "${PostgreSQL_INCLUDE_DIRS}"
--                                   INTERFACE_LINK_LIBRARIES
--                                   "${PostgreSQL_LIBRARIES}")
-+  target_link_libraries(pg_lib INTERFACE PostgreSQL::PostgreSQL)
-   mark_as_advanced(PG_INCLUDE_DIRS PG_LIBRARIES)
- endif(PostgreSQL_FOUND)
- include(FindPackageHandleStandardArgs)
+ if(BUILD_ORM)
+   # find postgres
+-  find_package(pg)
+-  if(pg_FOUND)
+-    message(STATUS "libpq inc path:" ${PG_INCLUDE_DIRS})
+-    message(STATUS "libpq lib:" ${PG_LIBRARIES})
+-    target_link_libraries(${PROJECT_NAME} PRIVATE pg_lib)
++  find_package(PostgreSQL)
++  if(PostgreSQL_FOUND)
++    message(STATUS "libpq inc path:" ${PostgreSQL_INCLUDE_DIRS})
++    message(STATUS "libpq lib:" ${PostgreSQL_LIBRARIES})
++    target_link_libraries(${PROJECT_NAME} PRIVATE ${PostgreSQL_LIBRARIES})
+     set(DROGON_SOURCES ${DROGON_SOURCES}
+                        orm_lib/src/postgresql_impl/PostgreSQLResultImpl.cc)
+     if(LIBPQ_BATCH_MODE)

